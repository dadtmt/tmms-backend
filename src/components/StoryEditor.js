import R from 'ramda'
import React, { Component, PropTypes } from 'react'
import { Button, ButtonToolbar, Checkbox, Col, Row } from 'react-bootstrap'

import {
  UPDATE_CHOICE_SUBSCRIPTION,
  UPDATE_TESTDICE_SUBSCRIPTION
} from '../graphql/subscriptions'
import CreateChoice from './CreateChoice'
import CreateTest from './CreateTest'
import CreateCrossroad from './CreateCrossroad'
import CurrentCrossroad from './CurrentCrossroad'
import Crossroads from './Crossroads'

export const getCrossroadsFromData = R.pipe(
  R.pathOr(
    [],
    ['viewer', 'user', 'editors', 'edges']
  ),
  R.head,
  R.pathOr([], ['node', 'crossroads', 'edges'])
)

export const getCurrentCrossroad = R.pipe(
  R.head,
  R.propOr(null, 'node')
)

export const isChoiceMade = R.pipe(
  R.converge(
    R.concat,
    [
      R.pipe(
        R.pathOr([], ['choices', 'edges']),
        R.map(R.path(['node', 'made']))
      ),
      R.pipe(
        R.pathOr([], ['testDices', 'edges']),
        R.map(R.path(['node', 'made']))
      )
    ]
  ),
  R.reduce(
    R.or,
    false
  )
)

class StoryEditor extends Component {
  constructor(props) {
    super(props)
    this.state = {
      showAddChoice: false,
      showAddTest: false
    }
  }

  componentWillReceiveProps(newProps) {
    const { data } = newProps
    const { loading } = data
    const crossroadId = R.propOr(
      null,
      'id'
    )(getCurrentCrossroad(getCrossroadsFromData(data)))
    const lastCrossroadId =
      R.propOr(
        null,
        'id'
      )(getCurrentCrossroad(getCrossroadsFromData(this.props.data)))

    if (!loading && !R.isNil(crossroadId) && crossroadId !== lastCrossroadId) {
      this.choiceSubscription = data.subscribeToMore({
        document: UPDATE_CHOICE_SUBSCRIPTION,
        updateQuery: prev => prev,
        variables: {
          choiceFilter: {
            crossroadId: {
              eq: crossroadId
            }
          }
        }
      })
      this.testDiceSubscription = data.subscribeToMore({
        document: UPDATE_TESTDICE_SUBSCRIPTION,
        updateQuery: prev => prev,
        variables: {
          testDiceFilter: {
            crossroadId: {
              eq: crossroadId
            }
          }
        }
      })
    }
  }

  render() {
    const {
      createChoice,
      deleteChoice,
      createCrossroad,
      createTest,
      deleteTestDice,
      data,
      toggleIsReady,
      updateCrossroadText
    } = this.props

    const { loading } = data

    const crossroads = getCrossroadsFromData(data)

    const currentCrossroad = getCurrentCrossroad(crossroads)

    const crossroadsWithoutCurrent = R.tail(crossroads)
    const allowNewCrossroad = R.isNil(currentCrossroad) ||
      isChoiceMade(currentCrossroad)

    const isCrossroadReady = R.propOr(false, 'isReady')(currentCrossroad)

    const crossroadId = R.propOr(null, 'id')(currentCrossroad)

    return (
      <Row>
        <Col sm={6}>
          {loading && <p>Loading...</p>}
          <CreateCrossroad
            createCrossroad={createCrossroad}
            crossroad={allowNewCrossroad ? null : currentCrossroad}
            updateCrossroadText={updateCrossroadText}
          />
          <Checkbox
            checked={isCrossroadReady}
            disabled={allowNewCrossroad}
            onChange={() => toggleIsReady({
              id: crossroadId,
              isReady: !isCrossroadReady
            })}
          >
            Make the page ready to play
          </Checkbox>
          <ButtonToolbar>
            <Button
              bsStyle='primary'
              disabled={allowNewCrossroad}
              onClick={() => this.setState({
                showAddChoice: true,
                showAddTest: false
              })}
            >
              Add a Choice
            </Button>
            <Button
              bsStyle='primary'
              disabled={allowNewCrossroad}
              onClick={() => this.setState({
                showAddChoice: false,
                showAddTest: true
              })}
            >
              Add a Test
            </Button>
          </ButtonToolbar>
          {(this.state.showAddChoice && !allowNewCrossroad) && <CreateChoice
            createChoice={createChoice}
            crossroadId={crossroadId}
          />}
          {(this.state.showAddTest && !allowNewCrossroad) && <CreateTest
            createTest={createTest}
            crossroadId={crossroadId}
          />}
        </Col>
        <Col sm={6}>
          {currentCrossroad && <CurrentCrossroad
            crossroad={currentCrossroad}
            deleteChoice={deleteChoice}
            deleteTestDice={deleteTestDice}
          />}
          <Crossroads crossroads={{ edges: crossroadsWithoutCurrent }} />
        </Col>
      </Row>
    )
  }
}

StoryEditor.propTypes = {
  createChoice: PropTypes.func.isRequired,
  createCrossroad: PropTypes.func.isRequired,
  createTest: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired,
  deleteChoice: PropTypes.func.isRequired,
  deleteTestDice: PropTypes.func.isRequired,
  toggleIsReady: PropTypes.func.isRequired,
  updateCrossroadText: PropTypes.func.isRequired
}

export default StoryEditor
