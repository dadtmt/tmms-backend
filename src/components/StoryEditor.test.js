import R from 'ramda'
import React from 'react'
import ReactDOM from 'react-dom'
import { ApolloProvider } from 'react-apollo'
import ApolloClient from 'apollo-client'

import StoryEditor, {
  getCrossroadsFromData,
  getCurrentCrossroad,
  isChoiceMade
} from './StoryEditor'

describe('getCurrentCrossroad', () => {
  it('should return null on loading', () => {
    const data = {
      loading: true
    }
    expect(getCurrentCrossroad(getCrossroadsFromData(data))).toMatchSnapshot()
  })
  it('should return current crossroad', () => {

    /* eslint-disable sort-keys */
    const data = {
      'variables': {
        'pageEditorId': 'UGFnZUVkaXRvcjox'
      },
      'loading': false,
      'networkStatus': 7,
      'viewer': {
        'user': {
          'editors': {
            'edges': [
              {
                'node': {
                  'id': 'UGFnZUVkaXRvcjox',
                  'crossroads': {
                    'edges': [
                      {
                        'node': {
                          'createdAt': '2017-03-07T11:37:25.000Z',
                          'id': 'Q3Jvc3Nyb2FkOjIx',
                          'isReady': false,
                          'text': {
                            'entityMap': {},
                            'blocks': [
                              {
                                'key': '8ussa',
                                'text': 'fsdfsfsdfsfd',
                                'type': 'unstyled',
                                'depth': 0,
                                'inlineStyleRanges': [],
                                'entityRanges': [],
                                'data': {}
                              }
                            ]
                          },
                          'choices': {
                            'edges': [],
                            '__typename': 'ChoiceConnection'
                          },
                          'testDices': {
                            'edges': [],
                            '__typename': 'TestDiceConnection'
                          },
                          '__typename': 'Crossroad'
                        },
                        '__typename': 'CrossroadEdge'
                      }
                    ],
                    '__typename': 'CrossroadConnection'
                  },
                  '__typename': 'PageEditor'
                },
                '__typename': 'UserEditorsEdge'
              }
            ],
            '__typename': 'UserEditorsConnection'
          },
          '__typename': 'User'
        },
        '__typename': 'Viewer'
      }
    }

    /* eslint-enable sort-keys */

    expect(getCurrentCrossroad(getCrossroadsFromData(data))).toMatchSnapshot()
  })
})

describe('isChoiceMade', () => {
  it('return true if one choice is made', () => {

    /* eslint-disable sort-keys */
    const data = {
      'variables': {
        'pageEditorId': 'UGFnZUVkaXRvcjox'
      },
      'loading': false,
      'networkStatus': 7,
      'viewer': {
        'user': {
          'editors': {
            'edges': [
              {
                'node': {
                  'id': 'UGFnZUVkaXRvcjox',
                  'crossroads': {
                    'edges': [
                      {
                        'node': {
                          'createdAt': '2017-03-07T11:37:25.000Z',
                          'id': 'Q3Jvc3Nyb2FkOjIx',
                          'isReady': false,
                          'text': {
                            'entityMap': {},
                            'blocks': [
                              {
                                'key': '8ussa',
                                'text': 'fsdfsfsdfsfd',
                                'type': 'unstyled',
                                'depth': 0,
                                'inlineStyleRanges': [],
                                'entityRanges': [],
                                'data': {}
                              }
                            ]
                          },
                          'choices': {
                            'edges': [
                              {
                                node: {
                                  made: true
                                }
                              },
                              {
                                node: {
                                  made: false
                                }
                              }
                            ],
                            '__typename': 'ChoiceConnection'
                          },
                          'testDices': {
                            'edges': [],
                            '__typename': 'TestDiceConnection'
                          },
                          '__typename': 'Crossroad'
                        },
                        '__typename': 'CrossroadEdge'
                      }
                    ],
                    '__typename': 'CrossroadConnection'
                  },
                  '__typename': 'PageEditor'
                },
                '__typename': 'UserEditorsEdge'
              }
            ],
            '__typename': 'UserEditorsConnection'
          },
          '__typename': 'User'
        },
        '__typename': 'Viewer'
      }
    }

    /* eslint-enable sort-keys */
    expect(
      isChoiceMade(getCurrentCrossroad(getCrossroadsFromData(data)))
    ).toBeTruthy()
  })

  it('return true if one test is made', () => {

    /* eslint-disable sort-keys */
    const data = {
      'variables': {
        'pageEditorId': 'UGFnZUVkaXRvcjox'
      },
      'loading': false,
      'networkStatus': 7,
      'viewer': {
        'user': {
          'editors': {
            'edges': [
              {
                'node': {
                  'id': 'UGFnZUVkaXRvcjox',
                  'crossroads': {
                    'edges': [
                      {
                        'node': {
                          'createdAt': '2017-03-07T11:37:25.000Z',
                          'id': 'Q3Jvc3Nyb2FkOjIx',
                          'isReady': false,
                          'text': {
                            'entityMap': {},
                            'blocks': [
                              {
                                'key': '8ussa',
                                'text': 'fsdfsfsdfsfd',
                                'type': 'unstyled',
                                'depth': 0,
                                'inlineStyleRanges': [],
                                'entityRanges': [],
                                'data': {}
                              }
                            ]
                          },
                          'choices': {
                            'edges': [
                              {
                                node: {
                                  made: false
                                }
                              },
                              {
                                node: {
                                  made: false
                                }
                              }
                            ],
                            '__typename': 'ChoiceConnection'
                          },
                          'testDices': {
                            'edges': [
                              {
                                node: {
                                  made: true
                                }
                              },
                              {
                                node: {
                                  made: false
                                }
                              }
                            ],
                            '__typename': 'TestDiceConnection'
                          },
                          '__typename': 'Crossroad'
                        },
                        '__typename': 'CrossroadEdge'
                      }
                    ],
                    '__typename': 'CrossroadConnection'
                  },
                  '__typename': 'PageEditor'
                },
                '__typename': 'UserEditorsEdge'
              }
            ],
            '__typename': 'UserEditorsConnection'
          },
          '__typename': 'User'
        },
        '__typename': 'Viewer'
      }
    }

    /* eslint-enable sort-keys */
    expect(
      isChoiceMade(getCurrentCrossroad(getCrossroadsFromData(data)))
    ).toBeTruthy()
  })
  it('return false if no crossroad', () => {
    const data = {
      loading: true
    }
    expect(
      isChoiceMade(getCurrentCrossroad(getCrossroadsFromData(data)))
    ).toBeFalsy()
  })
  it('return false if no choice is made', () => {

    /* eslint-disable sort-keys */
    const data = {
      'variables': {
        'pageEditorId': 'UGFnZUVkaXRvcjox'
      },
      'loading': false,
      'networkStatus': 7,
      'viewer': {
        'user': {
          'editors': {
            'edges': [
              {
                'node': {
                  'id': 'UGFnZUVkaXRvcjox',
                  'crossroads': {
                    'edges': [
                      {
                        'node': {
                          'createdAt': '2017-03-07T11:37:25.000Z',
                          'id': 'Q3Jvc3Nyb2FkOjIx',
                          'isReady': false,
                          'text': {
                            'entityMap': {},
                            'blocks': [
                              {
                                'key': '8ussa',
                                'text': 'fsdfsfsdfsfd',
                                'type': 'unstyled',
                                'depth': 0,
                                'inlineStyleRanges': [],
                                'entityRanges': [],
                                'data': {}
                              }
                            ]
                          },
                          'choices': {
                            'edges': [
                              {
                                node: {
                                  made: false
                                }
                              },
                              {
                                node: {
                                  made: false
                                }
                              }
                            ],
                            '__typename': 'ChoiceConnection'
                          },
                          'testDices': {
                            'edges': [
                              {
                                node: {
                                  made: false
                                }
                              },
                              {
                                node: {
                                  made: false
                                }
                              }
                            ],
                            '__typename': 'TestDiceConnection'
                          },
                          '__typename': 'Crossroad'
                        },
                        '__typename': 'CrossroadEdge'
                      }
                    ],
                    '__typename': 'CrossroadConnection'
                  },
                  '__typename': 'PageEditor'
                },
                '__typename': 'UserEditorsEdge'
              }
            ],
            '__typename': 'UserEditorsConnection'
          },
          '__typename': 'User'
        },
        '__typename': 'Viewer'
      }
    }

    /* eslint-enable sort-keys */

    expect(
      isChoiceMade(getCurrentCrossroad(getCrossroadsFromData(data)))
    ).toBeFalsy()
  })
})

it('renders without crashing', () => {
  const div = document.createElement('div')
  const props = {
    clearPageEditor: jest.fn(),
    createChoice: jest.fn(),
    createCrossroad: jest.fn(),
    createTest: jest.fn(),
    data: {},
    deleteChoice: jest.fn(),
    deleteTestDice: jest.fn(),
    toggleIsReady: jest.fn(),
    updateCrossroadText: jest.fn()
  }
  const client = new ApolloClient()
  ReactDOM.render(
    <ApolloProvider client={client}>
      <StoryEditor {...props} />
    </ApolloProvider>,
    div
  )
})
