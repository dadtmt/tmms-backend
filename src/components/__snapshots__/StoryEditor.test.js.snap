exports[`getCurrentCrossroad should return current crossroad 1`] = `
Object {
  "__typename": "Crossroad",
  "choices": Object {
    "__typename": "ChoiceConnection",
    "edges": Array [],
  },
  "createdAt": "2017-03-07T11:37:25.000Z",
  "id": "Q3Jvc3Nyb2FkOjIx",
  "isReady": false,
  "testDices": Object {
    "__typename": "TestDiceConnection",
    "edges": Array [],
  },
  "text": Object {
    "blocks": Array [
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "8ussa",
        "text": "fsdfsfsdfsfd",
        "type": "unstyled",
      },
    ],
    "entityMap": Object {},
  },
}
`;

exports[`getCurrentCrossroad should return null on loading 1`] = `null`;
